driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java

```diff
@@ -0,0 +1,303 @@
+import com.mongodb.MongoClientSettings;
+import com.mongodb.ReadConcern;
+import com.mongodb.ReadConcernLevel;
+import com.mongodb.TransactionOptions;
+import com.mongodb.WriteConcern;
+import com.mongodb.client.ClientSession;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.gridfs.GridFSBucket;
+import com.mongodb.client.gridfs.GridFSBuckets;
+import com.mongodb.internal.connection.TestCommandListener;
+import org.bson.BsonArray;
+import org.bson.BsonBoolean;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static com.mongodb.ClusterFixture.getMultiMongosConnectionString;
+import static com.mongodb.ClusterFixture.isSharded;
+import static com.mongodb.client.Fixture.getMongoClientSettingsBuilder;
+import static com.mongodb.client.unified.UnifiedCrudHelper.asReadConcern;
+import static com.mongodb.client.unified.UnifiedCrudHelper.asWriteConcern;
+import static java.util.Objects.requireNonNull;
+import static org.junit.Assume.assumeTrue;
+
+final class Entities {
+    private final Map<String, BsonValue> results = new HashMap<>();
+    private final Map<String, MongoClient> clients = new HashMap<>();
+    private final Map<String, MongoDatabase> databases = new HashMap<>();
+    private final Map<String, MongoCollection<BsonDocument>> collections = new HashMap<>();
+    private final Map<String, ClientSession> sessions = new HashMap<>();
+    private final Map<String, BsonDocument> sessionIdentifiers = new HashMap<>();
+    private final Map<String, GridFSBucket> buckets = new HashMap<>();
+    private final Map<String, TestCommandListener> clientCommandListeners = new HashMap<>();
+    private final Map<String, MongoCursor<BsonDocument>> changeStreams = new HashMap<>();
+
+    public void addResult(final String id, final BsonValue result) {
+        results.put(id, result);
+    }
+
+    public BsonValue getResult(final String id) {
+        BsonValue result = results.get(id);
+        if (result == null) {
+            throw new IllegalStateException("Missing change stream with id: " + id);
+        }
+        return result;
+    }
+
+    public void addChangeStream(final String id, final MongoCursor<BsonDocument> cursor) {
+        changeStreams.put(id, cursor);
+    }
+
+    public MongoCursor<BsonDocument> getChangeStream(final String id) {
+        MongoCursor<BsonDocument> changeStream = changeStreams.get(id);
+        if (changeStream == null) {
+            throw new IllegalStateException("Missing change stream with id: " + id);
+        }
+        return changeStream;
+    }
+
+    public boolean hasClient(final String id) {
+        return clients.containsKey(id);
+    }
+
+    public MongoClient getClient(final String id) {
```

Comment:
Maybe a nit pick, but what do you think about writing a method something like:

`
private static <T> T getItem(final String id, final HashMap<String, T> objects, final String name) {
        T object = objects.get(id);
        if (object == null) {
            throw new IllegalStateException("Missing " + name + "with id: " + id);
        }
        return object;
    }
`

Then each of these getters could be replaced with one liners like (I think the type can be inferred here no?):

`
return getItem(id, clients, "client");
`

Reply:
I like it! Done.

There is a lot of boilerplate in this PR and it's not clear how to get rid of most of it, but this is a start.